### ISSUE TITLE
Auto-expanding BufferManager needed

### ISSUE THREAD

**Issue by lefthandedgoat** [SuaveIO/suave#363]
I was doing some learning on the Http.fs 2.X and I kept crashing my suave server.

Server:
```
let webPart =
  choose
    [
      path "/" >>= choose [ GET >>= (OK "hello world") ]
    ]

startWebServer defaultConfig webPart
```

Crashing code:
```
open HttpFs.Client //this is the pre release v2.4 nuget
open System

let doit () =
    let request = createRequest Get (Uri("http://localhost:8083"))
    let response = getResponse request |> Async.RunSynchronously
    response.StatusCode
    
[| 1 .. 2000 |]
|> Array.iter (fun i -> printfn "%i %A" i (doit()))

System.Console.ReadKey() |> ignore
```

Its crashing because I am doing `let response` instead of `use response` so connections are being left open, but at the 50th connection attempt the exe hosting my suave server will crash.

On OSX+mono the app making the http requests and not closing them just hangs after 3-4 requests, but the suave server is still responding via a browser.

![image](https://cloud.githubusercontent.com/assets/725225/11980463/94fe191c-a960-11e5-8738-7c7c6b40803d.png)

**Comment by lefthandedgoat** [comment]
I don't know if its the same issue, but I was also able to crash it on OSX+Mono by simply issuing a whole bunch of concurrent requests.  I used the osx proxy called charles. Its like fiddler and will record your requests and you can repeat them.  It has an 'advanced repeat' option that allows you to issue X total requests with Y concurrency.  I figure anything over threadpool size of concurrency will push the suave server too hard.  Below is the stack I received.
```
Unhandled Exception:
System.InvalidOperationException: Stack empty.
  at System.ThrowHelper.ThrowInvalidOperationException (ExceptionResource resource) <0x47cd4a8 + 0x0002f> in <filename unknown>:0 
  at System.Collections.Generic.Stack`1[T].Pop () <0x2badb48 + 0x0001f> in <filename unknown>:0 
  at Suave.Sockets.BufferManager.PopBuffer (Microsoft.FSharp.Core.FSharpOption`1 context) <0x2bad980 + 0x0003b> in <filename unknown>:0 
  at Suave.Tcp+job@108.Invoke (Microsoft.FSharp.Core.Unit unitVar) <0x2bad3d8 + 0x001fb> in <filename unknown>:0 
  at Microsoft.FSharp.Control.AsyncBuilderImpl+callA@851[b,a].Invoke (Microsoft.FSharp.Control.AsyncParams`1 args) <0x7f56f0 + 0x000f8> in <filename unknown>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () <0x188f670 + 0x00035> in <filename unknown>:0 
  at Microsoft.FSharp.Control.CancellationTokenOps+Start@1268-1.Invoke (System.Runtime.ExceptionServices.ExceptionDispatchInfo edi) <0x47cd878 + 0x00013> in <filename unknown>:0 
  at Microsoft.FSharp.Control.AsyncBuilderImpl+callA@851[b,a].Invoke (Microsoft.FSharp.Control.AsyncParams`1 args) <0x7f56f0 + 0x00312> in <filename unknown>:0 
  at Microsoft.FSharp.Control.AsyncBuilderImpl+queueAsync@763[a].Invoke (Microsoft.FSharp.Core.Unit unitVar0) <0x7f4580 + 0x000d5> in <filename unknown>:0 
  at <StartupCode$FSharp-Core>.$Control.loop@430-52 (Microsoft.FSharp.Control.Trampoline this, Microsoft.FSharp.Core.FSharpFunc`2 action) <0x7f4538 + 0x0001f> in <filename unknown>:0 
  at Microsoft.FSharp.Control.Trampoline.ExecuteAction (Microsoft.FSharp.Core.FSharpFunc`2 firstAction) <0x7f4480 + 0x00053> in <filename unknown>:0 
  at Microsoft.FSharp.Control.TrampolineHolder.Protect (Microsoft.FSharp.Core.FSharpFunc`2 firstAction) <0x7f4330 + 0x0007b> in <filename unknown>:0 
  at <StartupCode$FSharp-Core>.$Control+-ctor@516-1.Invoke (System.Object state) <0x7f4188 + 0x00047> in <filename unknown>:0 
  at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context (System.Object state) <0x194eb90 + 0x0004f> in <filename unknown>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, Boolean preserveSyncCtx) <0x19487b0 + 0x001a4> in <filename unknown>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, Boolean preserveSyncCtx) <0x1948770 + 0x00033> in <filename unknown>:0 
  at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () <0x194eb00 + 0x0005a> in <filename unknown>:0 
  at System.Threading.ThreadPoolWorkQueue.Dispatch () <0x194d1c0 + 0x001d8> in <filename unknown>:0 
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () <0x194e970 + 0x00017> in <filename unknown>:0 

```

**Comment by ademar** [comment]
Just FYI this behaviour is partially by-design; if you want to support `n` clients concurrently you should set `maxOps = 2*n`

```fsharp
starWebServer {defaultConfig with maxOps = 1000 } (OK "yay")
```
In the future we may tackle this problem by having the buffer pool auto-size on demand.

Now, with respect to the first sample we should and we will implement a connection manager to expire those lingering connections and avoid getting DOSed.

**Comment by haf** [comment]
Marking this medium-hard because all one needs to do is write code similar to that in [EventStore/EventStore](https://github.com/EventStore/EventStore/blob/release-v3.5.0/src/EventStore.BufferManagement/BufferManager.cs), and the calls to allocate/release are already in the code, so only the implementation would have to be ported.

**Comment by haf** [comment]
Closing in favour of #380

**Comment by haf** [comment]
Fix released as v1.1
